tests/ondemand/ondemand_custom_types_document_tests.cpp

```diff
@@ -0,0 +1,177 @@
+    // which we expect to be faster.
+    for (auto field : obj) {
+      simdjson::ondemand::raw_json_string key;
+      if (auto const error = field.key().get(key)) {
+        return error;
+      }
+      if (key == "make") {
+        if (auto const error = field.value().get_string(car.make)) {
+          return error;
+        }
+      } else if (key == "model") {
+        if (auto const error = field.value().get_string(car.model)) {
+          return error;
+        }
+      } else if (key == "year") {
+        if (auto const error = field.value().get_int64().get(car.year)) {
+          return error;
+        }
+      } else if (key == "tire_pressure") {
+        if (auto const error = field.value().get<std::vector<double>>().get(
+                car.tire_pressure)) {
+          return error;
+        }
+      }
+    }
+    return car;
+  }
+};
+
+static_assert(simdjson::tag_invocable<simdjson::deserialize_tag,
+                                      std::type_identity<std::unique_ptr<Car>>,
+                                      simdjson::ondemand::value &>,
+              "It should be invocable");
+
+static_assert(simdjson::tag_invocable<simdjson::deserialize_tag,
+                                      std::type_identity<std::unique_ptr<Car>>,
+                                      simdjson::ondemand::document &>,
+              "Tag_invoke should work with document as well.");
+
+static_assert(simdjson::tag_invocable<simdjson::deserialize_tag,
+                                      std::type_identity<std::vector<Car>>,
+                                      simdjson::ondemand::value &>,
+              "It should be invocable");
+
+static_assert(simdjson::tag_invocable<simdjson::deserialize_tag,
+                                      std::type_identity<std::vector<Car>>,
+                                      simdjson::ondemand::document &>,
+              "Tag_invoke should work with document as well.");
+
+bool custom_test() {
+  TEST_START();
+  auto const json = R"( {
+      "make": "Toyota",
+      "model": "Camry",
+      "year": 2018,
+      "tire_pressure": [ 40.1, 39.9 ]
+ } )"_padded;
+
+  simdjson::ondemand::parser parser;
+  simdjson::ondemand::document doc = parser.iterate(json);
+
+  // creating car directly from "doc":
+  const std::unique_ptr<Car> car(doc);
+  if (car->make != "Toyota") {
+    return false;
+  }
+  if (car->model != "Camry") {
+    return false;
+  }
+  if (car->year != 2018) {
+    return false;
+  }
+  if (car->tire_pressure.size() != 2) {
+    return false;
+  }
+  return true;
+}
+
+bool custom_test_move_semantics() {
+  TEST_START();
+  auto const json = R"( {
+      "make": "Toyota",
+      "model": "Camry",
+      "year": 2018,
+      "tire_pressure": [ 40.1, 39.9 ]
+ } )"_padded;
+
+  simdjson::ondemand::parser parser;
+  simdjson::ondemand::document doc_orig = parser.iterate(json);
+  simdjson::ondemand::document&& doc = std::move(doc_orig);
```

Comment:
I see what you are doing here, but it is awfully complicated and people often look at the tests to see how to code.

It will almost never be the case that we will hold an rvalue so if we want to test these cases, I suggest we do it separately and not mess up the otherwise fine code.